File: Final_Task CPP/priority.cpp
#include "opencv2/opencv.hpp"
#include <iostream>
#include "global.hpp"
#include "mov.hpp"
#include "cam.hpp"
#include <sstream>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/videoio.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/video.hpp>

using namespace std;
using namespace cv;

Mat gframe;

int main()
{
	int priority = 1;
	int prev_priority = 1;

	int count = 0;
	int flag = 0;
	double eff = 0;

	int history = 45;
	float varThreshold = 200;
	bool bShadowDetection = true;

	string path = "/home/zetawolfx2/Desktop/Task_Video.mp4";
	VideoCapture cap(path);

	Ptr<BackgroundSubtractor> fgbg = createBackgroundSubtractorMOG2(history, varThreshold, bShadowDetection);

	while(1)
	{
		Mat frame;
		if(!(frame.dims>=2))
			cap >> frame;
		else
			break;

		int h = frame.size().height;
		int w = frame.size().width;

		Mat square = frame(Range((h/2)-70, (h/2)+120), Range((w/2)-155 ,(w/2)+155));
	
		Mat fgmask;
		fgbg->apply(square, fgmask);
	
		threshold(fgmask, fgmask, 30, 255, THRESH_BINARY);
    		Mat element = getStructuringElement(MORPH_RECT, Size(2, 2), Point(-1, -1));
		dilate(fgmask, fgmask, element, Point(-1, -1), 2);

		Mat grey;
		cvtColor(frame, grey, CV_BGR2GRAY);
		GaussianBlur(grey, grey, Size(3,3), 0, 0);
		Canny(grey, grey, 150, 250);

		Mat mask = Mat::zeros(grey.size().height, grey.size().width, CV_8U);
		Point poly_mask_points[1][4];
		poly_mask_points[0][0] = Point(50, h);
		poly_mask_points[0][1] = Point(1150, h);
		poly_mask_points[0][2] = Point(850, 480);
		poly_mask_points[0][3] = Point(450, 480);
		const Point* ppt[1] = { poly_mask_points[0] };
		int npt[] = { 4 };
		fillPoly(mask, ppt, npt, 1, Scalar( 255, 255, 255 ), LINE_8);
		Mat masked_img;
		bitwise_and(grey, mask, masked_img);

		Mat mask_left = Mat::zeros(grey.size().height, grey.size().width, CV_8U);
		Point poly_mask_left_points[1][4];
		poly_mask_left_points[0][0] = Point(50, h);
		poly_mask_left_points[0][1] = Point((w/2)-60, h);
		poly_mask_left_points[0][2] = Point((w/2)-60, 450);
		poly_mask_left_points[0][3] = Point(450, 450);
		const Point* ppt_l[1] = { poly_mask_left_points[0] };
		fillPoly(mask_left, ppt_l, npt, 1, Scalar( 255, 255, 255 ), LINE_8);
		Mat masked_img_left;
		bitwise_and(grey, mask_left, masked_img_left);

		Mat mask_right = Mat::zeros(grey.size().height, grey.size().width, CV_8U);
		Point poly_mask_right_points[1][4];
		poly_mask_right_points[0][0] = Point((w/2)+170, h);
		poly_mask_right_points[0][1] = Point(1150, h);
		poly_mask_right_points[0][2] = Point(850, 450);
		poly_mask_right_points[0][3] = Point((w/2)+170, 450);
		const Point* ppt_r[1] = { poly_mask_right_points[0] };
		fillPoly(mask_right, ppt_r, npt, 1, Scalar( 255, 255, 255 ), LINE_8);
		Mat masked_img_right;
		bitwise_and(grey, mask_right, masked_img_right);

		vector<Point> locations_left;
		findNonZero(masked_img_left, locations_left);
		Point l_pnt_l = locations_left[0];

		vector<Point> locations_right;
		findNonZero(masked_img_right, locations_right);
		Point l_pnt_r = locations_right[0];

		Point u_pnt_l = l_pnt_l;
		Point u_pnt_r = l_pnt_r;
		u_pnt_l.y -= 150;
		u_pnt_r.y -= 200;

		/*
		imshow("Mask", masked_img);
		imshow("LMask", masked_img_left);
		imshow("RMask", masked_img_right);
		imshow("Square", fgmask);
		imshow("Gray", grey);
		*/

		int h_ar = u_pnt_r.y - l_pnt_r.y;
		int w_ar = u_pnt_l.x - u_pnt_r.x;
		
		vector<Point> white;
		findNonZero(fgmask, white);
		double white_pixels = white.size();
		double ROI_Area = h_ar*w_ar;

		count = count + 1;

		if(count == 60)
		{
			eff = white_pixels/ROI_Area;
			eff = round( eff * 1000.0 ) / 1000.0;
			prev_priority = priority;
			if (eff<0.050)
				{
					priority = 1;
				}
			else if((eff>0.050)&&(eff<0.090))
			    	{
					priority = 2;
				}
			else if (eff>0.090)
			    	{
					priority = 3;
				}
			count = 0;
		}

		if(priority==1)
		{
			display_mov(1);
			display_cam(0);
		}
		else if ((priority == 2)&&(prev_priority == 1))
		{
			display_mov(1);
			display_cam(0);
			flag = 1;
		}
		else if ((priority == 2)&&(prev_priority == 2))
		{
			if (flag == 1)
			{
				display_cam(1);
				if (count == 59)
					flag = 0;
			}
			else if (flag == 0)
			{
				display_mov(1);
				display_cam(0);
				if (count == 59)
					flag = 1;
			}
		}
		else if(priority==3)
			display_cam(1);
		
		imshow("Stream", gframe);
		waitKey(15);
	}

	cap.release();

	destroyAllWindows();

	return 0;
		
}

File: final_task/scripts/priority.py
#!/usr/bin/env python

import cv2 as cv
import numpy as np
import rospy
from sensor_msgs.msg import Image
from final_task.srv import Flag, FlagResponse
from cv_bridge import CvBridge, CvBridgeError

def adjust_gamma(image, gamma=1.5):
    invGamma = 1.0 / gamma
    table = np.array([((i / 255.0) ** invGamma) * 255
        for i in np.arange(0, 256)]).astype("uint8")
    return cv.LUT(image, table)

path =  r'/home/zetawolfx2/catkin_ws/src/final_task/scripts/Task_Video.mp4'

rospy.init_node('central_priority_node', anonymous = True)

cap = cv.VideoCapture(path)

bridge = CvBridge()

fgbg = cv.createBackgroundSubtractorMOG2(history = 45, varThreshold = 200, detectShadows=True)

rospy.wait_for_service('camera_stream')
camera_stream = rospy.ServiceProxy('camera_stream', Flag)
rospy.wait_for_service('movie_stream')
movie_stream = rospy.ServiceProxy('movie_stream', Flag)

count = 0
flag = 0

prev_priority = 1
priority = 1

def callback(data):
	cv_image = bridge.imgmsg_to_cv2(data, desired_encoding = "bgr8")
	cv.imshow("Stream", cv_image)
	cv.waitKey(1)

rospy.Subscriber('movie_stream', Image, callback)
rospy.Subscriber('camera_stream', Image, callback)

while (cap.isOpened()):
	ret, frame = cap.read()
	copy = np.copy(frame)

	try:
		h, w = frame.shape[:-1]
	except:
		break

	square = frame[h//2-70:h//2+120, w//2-155:w//2+155]

	square = adjust_gamma(square)

	fgmask = fgbg.apply(square)

	ret, fgmask = cv.threshold(fgmask, 30, 255, cv.THRESH_BINARY)
	kernel = np.ones((2, 2), np.uint8)
	fgmask = cv.dilate(fgmask, kernel, iterations=2)

	grey = cv.cvtColor(frame, cv.COLOR_BGR2GRAY)

	hs, ws = square.shape[:-1]

	height, width = grey.shape

	blur = cv.GaussianBlur(grey, (5, 5), 1)

	edges = cv.Canny(grey, 150, 250)

	points = np.array([(50, height), (1150, height), (850, 480), (450, 480)])
	mask = np.zeros_like(edges)
	cv.fillPoly(mask, np.int32([points]), 255)
	masked_img = cv.bitwise_and(edges, mask)

	points1 = np.array([(width / 2 + 170, height), (1150, height), (850, 450), (width / 2 + 170, 450)])
	points2 = np.array([(50, height), (width / 2 - 60, height), (width / 2 - 60, 450), (450, 450)])
	mask_left = np.zeros_like(edges)
	cv.fillPoly(mask_left, np.int32([points2]), 255)
	mask_right = np.zeros_like(edges)
	cv.fillPoly(mask_right, np.int32([points1]), 255)
	masked_img_left = cv.bitwise_and(edges, mask_left)
	masked_img_right = cv.bitwise_and(edges, mask_right)

	indices_left = np.where(masked_img_left == [255])
	indices_right = np.where(masked_img_right == [255])

	point1 = (indices_left[1][0], indices_left[0][0])
	point2 = (indices_right[1][0], indices_right[0][0])
	point3 = (indices_left[1][0], 0)
	point4 = (indices_right[1][0], 0)
	point5 = (indices_left[1][0], indices_left[0][0] - 150)
	point6 = (indices_right[1][0], indices_right[0][0] - 200)

	ROI = frame[point6[1]:point2[1], point5[0]:point6[0]]

	indices = np.where(fgmask == [255])

	h_ar, w_ar = ROI.shape[:-1]

	ROI_Area = h_ar*w_ar

	count = count+1

	if(count==60):
		eff = float(ROI_Area/len(indices[0]))
		eff = 1/eff
		prev_priority = priority
		if (eff>0.090):
		    priority = 3
		elif (0.050<eff<0.090):
		    priority = 2
		elif (eff<0.050):
		    priority = 1
		count = 0

	if(priority==1):
		movie_stream(1)
		camera_stream(0)
	elif (priority == 2) and (prev_priority == 1):
		movie_stream(1)
		camera_stream(0)
		flag = 1
	elif (priority == 2) and (prev_priority == 2):
		if (flag == 1):
			movie_stream(0)
			camera_stream(1)
			if (count == 59):
				flag = 0
		elif (flag == 0):
			movie_stream(1)
			camera_stream(0)
			if (count == 59):
				flag = 1
	elif(priority==3):
		movie_stream(0)
		camera_stream(1)

	k = cv.waitKey(9) & 0xff
	if k == 27:
		break

cap.release()
cv.destroyAllWindows()
